### PR TITLE
fix: Handle selecting files and folders for patch options correctly

### LINKED ISSUE #1703: bug: Can't select file paths
### Bug description

On a Samsung phone:
1. Launch Revanced Manager 
2. Patcher 
3. Select an application
4. Storage
5. Recommended version of youtube apk
6. Selected Patches
7. Custom Branding
8. Add options
9. App icon
10. Custom value
11. 3 dots next to ReVanced*logo
12. Select folder
13. Go to the folder that contain required folders with files
14. Tap on "use this folder"
15. Tap on authorize
16. Start patching
17. There's an error which says "No such file or directory" when revanced tries to eccess the given folder

### Error logs

```shell
- Device Info
ReVanced Manager: 1.18.0
Model: SM-G991B
Android version: 14
Supported architectures: arm64-v8a, armeabi-v7a, armeabi
Root permissions: No

- Patch Info
App: com.google.android.youtube v19.04.37 (Suggested: v19.04.37)
Patches version: v4.2.0
Patches added: Change header, Custom branding [App name: YT Revanced, App icon: content://com.android.externalstorage.documents/tree/primary%3ARevanced%2FRevanced%20Icon%20Pack]
Patches removed: None
Default patch options changed: None

- Settings
Allow changing patch selection: true
Version compatibility check: true
Show universal patches: false
Patches source: revanced/revanced-patches
Integration source: revanced/revanced-integrations

- Logs
Copying APK
Reading APK
Decoding app manifest
Loading patches
Merging integrations
Deleting existing resource cache directory
Decoding resources
Executing patches
Applied 62 patches
Custom branding failed: app.revanced.patcher.patch.PatchException: content:/com.android.externalstorage.documents/tree/primary%3ARevanced%2FRevanced%20Icon%20Pack/mipmap-xxxhdpi/adaptiveproduct_youtube_background_color_108.png: open failed: ENOENT (No such file or directory)
    at app.revanced.patcher.Patcher$apply$1.invokeSuspend$executePatch(Patcher.kt:185)
    at app.revanced.patcher.Patcher$apply$1.invokeSuspend(Patcher.kt:203)
    at app.revanced.patcher.Patcher$apply$1.invoke(Unknown Source:8)
    at app.revanced.patcher.Patcher$apply$1.invoke(Unknown Source:4)
    at kotlinx.coroutines.flow.SafeFlow.collectSafely(Builders.kt:61)
    at kotlinx.coroutines.flow.AbstractFlow.collect(Flow.kt:230)
    at app.revanced.manager.flutter.MainActivity$runPatcher$1$2$1.invokeSuspend(MainActivity.kt:341)
    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
    at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:280)
    at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
    at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source:1)
    at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
    at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source:1)
    at app.revanced.manager.flutter.MainActivity.runPatcher$lambda$36(MainActivity.kt:340)
    at app.revanced.manager.flutter.MainActivity.$r8$lambda$wwTqh4bVDMuwAT_qXBREk8gLHGU(Unknown Source:0)
    at app.revanced.manager.flutter.MainActivity$$ExternalSyntheticLambda5.run(Unknown Source:24)
    at java.lang.Thread.run(Thread.java:1012)
Caused by: java.io.FileNotFoundException: content:/com.android.externalstorage.documents/tree/primary%3ARevanced%2FRevanced%20Icon%20Pack/mipmap-xxxhdpi/adaptiveproduct_youtube_background_color_108.png: open failed: ENOENT (No such file or directory)
    at libcore.io.IoBridge.open(IoBridge.java:574)
    at java.io.FileInputStream.<init>(FileInputStream.java:160)
    at kotlin.io.FilesKt__FileReadWriteKt.readBytes(FileReadWrite.kt:63)
    at app.revanced.patches.youtube.layout.branding.CustomBrandingPatch.execute(CustomBrandingPatch.kt:90)
    at app.revanced.patches.youtube.layout.branding.CustomBrandingPatch.execute(CustomBrandingPatch.kt:14)
    at app.revanced.patcher.Patcher$apply$1.invokeSuspend$executePatch(Patcher.kt:177)
    ... 18 more
Caused by: android.system.ErrnoException: open failed: ENOENT (No such file or directory)
    at libcore.io.Linux.open(Native Method)
    at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
    at libcore.io.BlockGuardOs.open(BlockGuardOs.java:274)
    at libcore.io.ForwardingOs.open(ForwardingOs.java:563)
    at android.app.ActivityThread$AndroidOs.open(ActivityThread.java:8756)
    at libcore.io.IoBridge.open(IoBridge.java:560)
    ... 23 more
Compiling patched dex files
Compiled 8 dex files
Compiling modified resources
Signing APK
Patched APK
```

### Solution

Maybe the authorize button doesn't work ? Revanced doesn't let me chose an other file explorer...

### Additional context

_No response_

### Acknowledgements

- [X] This issue is not a duplicate of an existing bug report.
- [X] I have chosen an appropriate title.
- [X] All requested information has been provided properly.

### PR DESCRIPTION
**Pull request by Francesco146**
fixes #1703

### CODE REVIEW

**Comment by Francesco146** [comment]
ops, i opened the PR for the main branch, not sure what will happen if i change it to the dev branch

File: lib/ui/widgets/patchesSelectorView/patch_options_fields.dart [hunk 1]
```diff
@@ -1,6 +1,8 @@
+import 'package:file_picker/file_picker.dart';
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
 import 'package:flutter_file_dialog/flutter_file_dialog.dart';
```
Comment (Francesco146) [inline lib/ui/widgets/patchesSelectorView/patch_options_fields.dart:4]: waiting for the possible branch switch, this import is not needed anymore

File: lib/ui/widgets/patchesSelectorView/patch_options_fields.dart [hunk 2]
```diff
@@ -503,25 +505,44 @@ class _TextFieldForPatchOptionState extends State<TextFieldForPatchOption> {
...
-                        widget.onChanged(controller.text);
+                  // manageExternalStorage permission is required for file/folder selection
+                  // otherwise, the app will not complain, but the patches will error out
+                  // the same way as if the user selected an empty file/folder
+                  final Map<String, dynamic> availableActions = {
+                    t.patchOptionsView.selectFilePath: () async {
+                      final FilePickerResult? result =
+                          await FilePicker.platform.pickFiles();
+                      final permission =
+                          await Permission.manageExternalStorage.request();
```
Comment (Francesco146) [inline lib/ui/widgets/patchesSelectorView/patch_options_fields.dart:516]: here maybe request the permission only if the platform is Android 11+

**Comment by Francesco146** [comment]
ready for review

**Comment by Francesco146** [comment]
update:

this PR has been merged in RVX Manager, many people have already tested it and it has no problems

File: lib/ui/widgets/patchesSelectorView/patch_options_fields.dart [hunk 3]
```diff
@@ -1,10 +1,17 @@
...
 import 'package:flutter/material.dart';
 import 'package:flutter/services.dart';
-import 'package:flutter_file_dialog/flutter_file_dialog.dart';
+import 'package:permission_handler/permission_handler.dart';
 import 'package:revanced_manager/gen/strings.g.dart';
 import 'package:revanced_manager/models/patch.dart';
 import 'package:revanced_manager/ui/widgets/shared/custom_card.dart';
 
+import '../../../app/app.locator.dart';
+import '../../../services/manager_api.dart';
```
Comment (validcube) [inline lib/ui/widgets/patchesSelectorView/patch_options_fields.dart:13]: 🔥 Avoid using relative import.

```suggestion
import 'package:revanced_manager/app/app.locator.dart';
import 'package:revanced_manager/services/manager_api.dart';
```

File: lib/ui/widgets/patchesSelectorView/patch_options_fields.dart [hunk 4]
```diff
@@ -1,10 +1,17 @@
+import 'dart:io';
+
+import 'package:device_info_plus/device_info_plus.dart';
```
Comment (validcube) [inline lib/ui/widgets/patchesSelectorView/patch_options_fields.dart:3]: 🤔 These appears to be unused.

```suggestion
```

**Review by validcube — CHANGES_REQUESTED** [review]
📝 Review: Analysis & Formating

(Note-to-self: Revisit this to test on physical devices when have time.)

**Comment by Francesco146** [comment]
> (Note-to-self: Revisit this to test on physical devices when have time.)

i tested the latest commit with an Android 14 virtual emulator, 10+ should be covered. consider testing on Android 9

**Comment by Francesco146** [comment]
tested on AVD Android 9. working as expected

**Comment by kitadai31** [comment]
I tried this but it's currently broken...
There are mistakes when resolved a conflict

5c68d51 added a [new import](https://github.com/ReVanced/revanced-manager/commit/5c68d513a38dcca1b884429c28f5bb32697bf37d#diff-6575bf694d851412b87d3899a829b95c173207ab57d4f9b239597795923f41fcR6), but it was deleted by this merge, so build fails.

Also, 5c68d51 changed a [value strings of PopupMenuItem](https://github.com/ReVanced/revanced-manager/commit/5c68d513a38dcca1b884429c28f5bb32697bf37d#diff-6575bf694d851412b87d3899a829b95c173207ab57d4f9b239597795923f41fcL505-L516), but this change was [lost](https://github.com/ReVanced/revanced-manager/pull/1941/files#diff-6575bf694d851412b87d3899a829b95c173207ab57d4f9b239597795923f41fcR535) while resolving conflict.
Therefore, when I press "Select file path" or "Select folder", nothing happens.

**Comment by Francesco146** [comment]
should I open another PR? conflict resolution was not included in the goal of this PR

**Comment by Ushie** [comment]
Yes, open one please

**Comment by Francesco146** [comment]
@oSumAtrIX this was reverted in 1.21

**Comment by oSumAtrIX** [comment]
Another commit supposedly fixed it according to @TheAabedKhan

**Comment by Francesco146** [comment]
doesn't seem like so on my part

**Comment by oSumAtrIX** [comment]
Open a new issue

**Comment by Francesco146** [comment]
https://github.com/ReVanced/revanced-manager/issues/1703

there's already an issue open for this, basically this implementation gets a path as a content URI, that doesn't work with the patcher. the dependency that I added opens the system file/folder picker and returns the absolute path of what was selected. absolute paths works with the patcher and solves the missing permissions that A11+ requires

**Comment by oSumAtrIX** [comment]
The PR was reverted because of it breaking entering null in patch options:

https://github.com/ReVanced/revanced-manager/pull/1941/files#diff-6575bf694d851412b87d3899a829b95c173207ab57d4f9b239597795923f41fcL545-L549

**Comment by Francesco146** [comment]
okay, I'll create another pr that handles that? or someone is already working on it?

**Comment by oSumAtrIX** [comment]
The linked issue is not assigned to anyone yet, so sure.

**Pull request merged**